modules/component/src/state/state.ts

```diff
@@ -0,0 +1,249 @@
+  select<K1 extends keyof T, K2 extends keyof T[K1]>(
+    k1: K1,
+    k2: K2
+  ): Observable<T[K1][K2]>;
+  select<
+    K1 extends keyof T,
+    K2 extends keyof T[K1],
+    K3 extends keyof T[K1][K2]
+  >(k1: K1, k2: K2, k3: K3): Observable<T[K1][K2][K3]>;
+  select<
+    K1 extends keyof T,
+    K2 extends keyof T[K1],
+    K3 extends keyof T[K1][K2],
+    K4 extends keyof T[K1][K2][K3]
+  >(k1: K1, k2: K2, k3: K3, k4: K4): Observable<T[K1][K2][K3][K4]>;
+  select<
+    K1 extends keyof T,
+    K2 extends keyof T[K1],
+    K3 extends keyof T[K1][K2],
+    K4 extends keyof T[K1][K2][K3],
+    K5 extends keyof T[K1][K2][K3][K4]
+  >(k1: K1, k2: K2, k3: K3, k4: K4, k5: K5): Observable<T[K1][K2][K3][K4][K5]>;
+  select<
+    K1 extends keyof T,
+    K2 extends keyof T[K1],
+    K3 extends keyof T[K1][K2],
+    K4 extends keyof T[K1][K2][K3],
+    K5 extends keyof T[K1][K2][K3][K4],
+    K6 extends keyof T[K1][K2][K3][K4][K5]
+  >(
+    k1: K1,
+    k2: K2,
+    k3: K3,
+    k4: K4,
+    k5: K5,
+    k6: K6
+  ): Observable<T[K1][K2][K3][K4][K5][K6]>;
+  // ===========================
+  select(...opOrMapFn: any[]): Observable<any> {
+    if (!opOrMapFn || opOrMapFn.length === 0) {
+      return this.state$.pipe(stateful());
+    } else if (!this.isStringArray(opOrMapFn)) {
+      const path = (opOrMapFn as any) as string[];
+      return this.state$.pipe(
+        pluck(...path),
+        stateful()
+      );
+    } else if (this.isOperateFnArray(opOrMapFn)) {
+      const oprs = opOrMapFn as OperatorFunction<T, any>[];
+      return this.state$.pipe(
+        pipeFromArray(oprs),
+        stateful()
+      );
+    }
+
+    throw new Error(
+      'Wrong params passed! Passed param:' + JSON.stringify(opOrMapFn)
+    );
+  }
+
+  hold<S>(observableWithSideEffect: Observable<S>): void;
+  hold<S>(
+    obsOrObsWithSideEffect: Observable<S>,
+    sideEffectFn?: (arg: S) => void
+  ): void {
+    if (sideEffectFn) {
+      this.effectSubject.next(obsOrObsWithSideEffect.pipe(tap(sideEffectFn)));
+    }
+    this.effectSubject.next(obsOrObsWithSideEffect);
+  }
+
+  /**
+   * teardown(): void
+   *
+   * When called it teardown all internal logic
+   * used to connect to the `OnDestroy` life-cycle hook of services, components, directives, pipes
+   */
+  teardown(): void {
+    this.subscription.unsubscribe();
+  }
+
+  private isOperateFnArray(op: any[]): op is OperatorFunction<T, any>[] {
+    return op.every((i: any) => typeof i !== 'string');
+  }
+
+  private isStringArray(op: any[]): op is string[] {
+    return op.every((i: any) => typeof i !== 'string');
+  }
+
+  /**
+   * ngOnDestroy(): void
+   *
+   * When called it teardown all internal logic
+   * used to connect to the `OnDestroy` life-cycle hook of services, components, directives, pipes
+   */
+  ngOnDestroy(): void {
+    this.teardown();
+  }
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export class GlobalState<T> extends State<T> {}
```

Comment:
Move this into another file, otherwise bazel won't be able to build it. 
(don't forget to add it to the `index.ts` file afterwards 🙂)